implementations/rust/ockam/ockam/src/relay_service/relay_service.rs

```diff
@@ -46,34 +61,75 @@ impl RelayService {
 #[crate::worker]
 impl Worker for RelayService {
     type Context = Context;
-    type Message = Any;
+    type Message = String;
 
     async fn handle_message(
         &mut self,
         ctx: &mut Self::Context,
-        msg: Routed<Self::Message>,
+        message: Routed<Self::Message>,
     ) -> Result<()> {
-        let forward_route = msg.return_route();
-        let payload = msg.payload();
-
-        let random_address = Address::random_tagged("Relay.service");
-
-        // TODO: assume that the first byte is length, ignore it.
-        // We have to improve this actually parse the payload.
-        let address = match payload.get(1..) {
-            Some(address) => match from_utf8(address) {
-                Ok(v) if v != "register" => Address::from_string(v),
-                _ => random_address,
-            },
-            None => random_address,
+        let secure_channel_local_info =
+            IdentitySecureChannelLocalInfo::find_info(message.local_message()).ok();
+
+        let forward_route = message.return_route();
+        let requested_relay_address = message.into_body()?;
+
+        let requested_relay_name = if requested_relay_address == "register" {
+            Address::random_tagged("Relay.service")
+                .address()
+                .to_string()
+        } else {
+            requested_relay_address
         };
 
+        debug!(%requested_relay_name, "Relay creation request");
+
+        // Verify the relay usage only when an authority is set, otherwise allow any relay name
+        if let Some(authority_validation) = &self.options.authority_validation {
+            if let Some(secure_channel_local_info) = secure_channel_local_info {
+                let attributes = authority_validation
+                    .identities_attributes
+                    .get_attributes(
+                        &secure_channel_local_info.their_identity_id(),
+                        &authority_validation.authority,
+                    )
+                    .await?;
+
+                if let Some(attributes) = attributes
+                    .and_then(|a| a.attrs().get("ockam-relay".as_bytes()).cloned())
+                    .and_then(|a| String::from_utf8(a).ok())
+                {
+                    match attributes.as_str() {
+                        "*" => {
+                            // allow any relay name
+                        }
+                        allowed_name => {
+                            if allowed_name != requested_relay_name {
+                                warn!(%allowed_name, "Relay creation request not authorized, relay name does not match the attribute, dropping.");
```

Comment:
It could be a good idea to include the requested name in order to better diagnose the issue, if there is a configuration problem.